File: scraper/test_hec_scraper.py
"""
Test script to inspect HEC website structure before full scraping
"""

import time
import os
import sys
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.chrome.options import Options

# Try to import webdriver_manager, but handle if it fails
try:
    from webdriver_manager.chrome import ChromeDriverManager
    WEBDRIVER_MANAGER_AVAILABLE = True
except ImportError:
    WEBDRIVER_MANAGER_AVAILABLE = False
    print("⚠️  webdriver-manager not available, will try alternative methods")


def setup_driver():
    """Setup Chrome driver with multiple fallback methods"""
    chrome_options = Options()
    chrome_options.add_argument('--no-sandbox')
    chrome_options.add_argument('--disable-dev-shm-usage')
    chrome_options.add_argument('--disable-blink-features=AutomationControlled')
    chrome_options.add_argument('--user-agent=Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36')
    chrome_options.add_experimental_option("excludeSwitches", ["enable-automation"])
    chrome_options.add_experimental_option('useAutomationExtension', False)
    
    driver = None
    
    # Method 1: Try using Chrome's built-in driver (Chrome 115+)
    try:
        print("   Trying Chrome's built-in driver...")
        driver = webdriver.Chrome(options=chrome_options)
        print("   ✅ Success with built-in driver!")
        driver.maximize_window()
        return driver
    except Exception as e:
        print(f"   ⚠️  Built-in driver failed: {e}")
    
    # Method 2: Try webdriver-manager
    if WEBDRIVER_MANAGER_AVAILABLE:
        try:
            print("   Trying webdriver-manager...")
            service = Service(ChromeDriverManager().install())
            driver = webdriver.Chrome(service=service, options=chrome_options)
            print("   ✅ Success with webdriver-manager!")
            driver.maximize_window()
            return driver
        except Exception as e:
            print(f"   ⚠️  webdriver-manager failed: {e}")
    
    # Method 3: Try common ChromeDriver paths
    common_paths = [
        r'C:\Program Files\Google\Chrome\Application\chromedriver.exe',
        r'C:\Program Files (x86)\Google\Chrome\Application\chromedriver.exe',
        os.path.join(os.getcwd(), 'chromedriver.exe'),
        os.path.join(os.path.expanduser('~'), 'chromedriver.exe')
    ]
    
    for path in common_paths:
        if os.path.exists(path):
            try:
                print(f"   Trying ChromeDriver at: {path}")
                service = Service(path)
                driver = webdriver.Chrome(service=service, options=chrome_options)
                print("   ✅ Success with local ChromeDriver!")
                driver.maximize_window()
                return driver
            except Exception as e:
                print(f"   ⚠️  Failed with {path}: {e}")
                continue
    
    # If all methods fail, raise error with instructions
    raise Exception("""
❌ Could not setup ChromeDriver!

Please try one of these solutions:

1. Update Chrome browser to latest version (Chrome 115+ has built-in driver)
   Download: https://www.google.com/chrome/

2. Manually download ChromeDriver:
   - Go to: https://chromedriver.chromium.org/downloads
   - Download matching your Chrome version
   - Extract chromedriver.exe to scraper folder
   - Run script again

3. Install ChromeDriver via Chocolatey (if you have it):
   choco install chromedriver

4. Check Chrome version:
   - Open Chrome
   - Go to: chrome://version/
   - Note the version number
   - Download matching ChromeDriver
""")


def main():
    print("🔍 Testing HEC Website Structure...\n")
    
    driver = setup_driver()
    
    try:
        print("📄 Navigating to HEC website...")
        print("⏳ This may take 30-60 seconds, please wait...\n")
        
        driver.get('https://www.hec.gov.pk/english/universities/Pages/recognised.aspx')
        time.sleep(8)  # Wait for page to load
        
        print("✅ Page loaded!\n")
        print("📊 Inspecting Page Structure...\n")
        
        # Find all select elements
        selects = driver.find_elements(By.TAG_NAME, 'select')
        print(f"🔽 Found {len(selects)} select elements:\n")
        
        for idx, select in enumerate(selects, 1):
            select_id = select.get_attribute('id') or 'no-id'
            select_name = select.get_attribute('name') or 'no-name'
            select_class = select.get_attribute('class') or 'no-class'
            
            print(f"   Select #{idx}:")
            print(f"      ID: {select_id}")
            print(f"      Name: {select_name}")
            print(f"      Class: {select_class}")
            
            # Get options
            options = select.find_elements(By.TAG_NAME, 'option')
            print(f"      Options ({len(options)} total):")
            for opt in options[:10]:  # First 10 options
                print(f"         - {opt.text.strip()}")
            if len(options) > 10:
                print(f"         ... and {len(options) - 10} more")
            print()
        
        # Find university elements
        print("🎓 Testing University Extraction...\n")
        
        # Try different selectors
        selectors = [
            ("//div[contains(@class, 'university')]", "University class"),
            ("//div[contains(@class, 'card')]", "Card class"),
            ("//li[contains(@class, 'university')]", "University li"),
            ("//div[contains(text(), 'University')]", "Text contains University")
        ]
        
        for xpath, description in selectors:
            try:
                elements = driver.find_elements(By.XPATH, xpath)
                if elements:
                    print(f"   ✅ {description}: Found {len(elements)} elements")
                    # Show first 3
                    for i, el in enumerate(elements[:3], 1):
                        text = el.text.strip()[:80]
                        print(f"      {i}. {text}...")
                else:
                    print(f"   ❌ {description}: No elements found")
            except Exception as e:
                print(f"   ⚠️  {description}: Error - {e}")
        
        print("\n\n✅ Test Complete!")
        print("📝 Review the output above to understand the page structure.")
        print("💡 If filters are detected, you can proceed with full scraping.\n")
        
        # Keep browser open for inspection
        print("⏳ Keeping browser open for 15 seconds for inspection...")
        time.sleep(15)
        
    except Exception as e:
        print(f"\n❌ Test Error: {e}")
        import traceback
        traceback.print_exc()
        
    finally:
        driver.quit()
        print("\n👋 Browser closed.")


if __name__ == '__main__':
    main()

File: backend/scripts/testHECScraper.js
const puppeteer = require('puppeteer');

/**
 * Test script to inspect HEC website structure
 * This will help us understand the actual HTML structure before full scraping
 */
async function testHECStructure() {
  console.log('🔍 Testing HEC Website Structure...\n');
  
  const browser = await puppeteer.launch({
    headless: false, // Keep visible to see what's happening
    args: [
      '--no-sandbox', 
      '--disable-setuid-sandbox',
      '--disable-blink-features=AutomationControlled',
      '--disable-dev-shm-usage'
    ]
  });

  try {
    const page = await browser.newPage();
    
    // Set user agent to avoid detection
    await page.setUserAgent('Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36');
    
    await page.setViewport({ width: 1920, height: 1080 });
    
    console.log('📄 Navigating to HEC website...');
    console.log('⏳ This may take 30-60 seconds, please wait...\n');
    
    // Try multiple navigation strategies
    let navigationSuccess = false;
    const strategies = [
      { waitUntil: 'load', timeout: 120000 },
      { waitUntil: 'domcontentloaded', timeout: 120000 },
      { waitUntil: 'networkidle0', timeout: 90000 }
    ];

    for (let i = 0; i < strategies.length; i++) {
      try {
        console.log(`   Trying strategy ${i + 1}/3: ${strategies[i].waitUntil}...`);
        await page.goto('https://www.hec.gov.pk/english/universities/Pages/recognised.aspx', strategies[i]);
        navigationSuccess = true;
        console.log('   ✅ Page loaded successfully!\n');
        break;
      } catch (error) {
        console.log(`   ⚠️  Strategy ${i + 1} failed: ${error.message}`);
        if (i < strategies.length - 1) {
          console.log('   Retrying with different strategy...\n');
          await page.waitForTimeout(2000);
        }
      }
    }

    if (!navigationSuccess) {
      throw new Error('Failed to load page with all strategies. Please check your internet connection.');
    }

    console.log('⏳ Waiting for page content to fully render...');
    await page.waitForTimeout(8000); // Give extra time for dynamic content

    // Inspect page structure
    console.log('\n📊 Inspecting Page Structure...\n');
    
    const pageInfo = await page.evaluate(() => {
      const info = {
        title: document.title,
        url: window.location.href,
        filters: {},
        universities: [],
        selectors: {}
      };

      // Find all select elements
      const selects = Array.from(document.querySelectorAll('select'));
      info.selects = selects.map((sel, idx) => ({
        index: idx,
        name: sel.name || 'no-name',
        id: sel.id || 'no-id',
        className: sel.className || 'no-class',
        options: Array.from(sel.options).map(opt => ({
          text: opt.text.trim(),
          value: opt.value
        })).slice(0, 10) // First 10 options
      }));

      // Find filter labels
      const labels = Array.from(document.querySelectorAll('label, span, div')).filter(el => {
        const text = el.textContent?.toLowerCase() || '';
        return text.includes('sector') || 
               text.includes('chartered') || 
               text.includes('discipline') || 
               text.includes('province') || 
               text.includes('city');
      });

      info.filterLabels = labels.slice(0, 10).map(el => ({
        tag: el.tagName,
        text: el.textContent?.trim().substring(0, 50),
        className: el.className,
        id: el.id
      }));

      // Find university cards/elements
      const uniElements = Array.from(document.querySelectorAll('div, li, article, section')).filter(el => {
        const text = el.textContent || '';
        return text.includes('University') || text.includes('Institute');
      });

      info.universityElements = uniElements.slice(0, 5).map(el => ({
        tag: el.tagName,
        className: el.className,
        id: el.id,
        text: el.textContent?.trim().substring(0, 100)
      }));

      // Find all classes that might be related to filters
      const allClasses = Array.from(document.querySelectorAll('*')).map(el => el.className)
        .filter(cls => cls && typeof cls === 'string')
        .filter(cls => cls.includes('filter') || cls.includes('select') || cls.includes('dropdown'))
        .slice(0, 20);

      info.filterClasses = [...new Set(allClasses)];

      return info;
    });

    // Print results
    console.log('📋 Page Information:');
    console.log(`   Title: ${pageInfo.title}`);
    console.log(`   URL: ${pageInfo.url}\n`);

    console.log('🔽 Select Elements Found:');
    pageInfo.selects.forEach((sel, idx) => {
      console.log(`\n   Select #${idx + 1}:`);
      console.log(`      Name: ${sel.name}`);
      console.log(`      ID: ${sel.id}`);
      console.log(`      Class: ${sel.className}`);
      console.log(`      Options (first 10):`);
      sel.options.forEach(opt => {
        console.log(`         - ${opt.text} (value: ${opt.value})`);
      });
    });

    console.log('\n🏷️  Filter Labels Found:');
    pageInfo.filterLabels.forEach((label, idx) => {
      console.log(`   ${idx + 1}. [${label.tag}] ${label.text}`);
      console.log(`      Class: ${label.className}, ID: ${label.id}`);
    });

    console.log('\n🎓 University Elements Found:');
    pageInfo.universityElements.forEach((uni, idx) => {
      console.log(`   ${idx + 1}. [${uni.tag}] ${uni.text}`);
      console.log(`      Class: ${uni.className}, ID: ${uni.id}`);
    });

    console.log('\n🎨 Filter-Related Classes:');
    pageInfo.filterClasses.forEach(cls => {
      console.log(`   - ${cls}`);
    });

    // Try to extract filter options
    console.log('\n\n🔍 Attempting to Extract Filter Options...\n');
    
    const filterOptions = await page.evaluate(() => {
      const filters = {};
      
      // Try to find selects by common patterns
      const sectorSelect = Array.from(document.querySelectorAll('select')).find(s => 
        s.name?.toLowerCase().includes('sector') || 
        s.id?.toLowerCase().includes('sector') ||
        s.previousElementSibling?.textContent?.toLowerCase().includes('sector')
      );
      
      if (sectorSelect) {
        filters.sector = Array.from(sectorSelect.options).map(opt => opt.text.trim());
      }

      const charteredSelect = Array.from(document.querySelectorAll('select')).find(s => 
        s.name?.toLowerCase().includes('chartered') || 
        s.id?.toLowerCase().includes('chartered')
      );
      
      if (charteredSelect) {
        filters.charteredBy = Array.from(charteredSelect.options).map(opt => opt.text.trim());
      }

      const disciplineSelect = Array.from(document.querySelectorAll('select')).find(s => 
        s.name?.toLowerCase().includes('discipline') || 
        s.id?.toLowerCase().includes('discipline')
      );
      
      if (disciplineSelect) {
        filters.discipline = Array.from(disciplineSelect.options).map(opt => opt.text.trim());
      }

      const provinceSelect = Array.from(document.querySelectorAll('select')).find(s => 
        s.name?.toLowerCase().includes('province') || 
        s.id?.toLowerCase().includes('province')
      );
      
      if (provinceSelect) {
        filters.province = Array.from(provinceSelect.options).map(opt => opt.text.trim());
      }

      const citySelect = Array.from(document.querySelectorAll('select')).find(s => 
        s.name?.toLowerCase().includes('city') || 
        s.id?.toLowerCase().includes('city')
      );
      
      if (citySelect) {
        filters.city = Array.from(citySelect.options).map(opt => opt.text.trim());
      }

      return filters;
    });

    console.log('📊 Extracted Filter Options:');
    Object.entries(filterOptions).forEach(([key, values]) => {
      if (values && values.length > 0) {
        console.log(`\n   ${key.toUpperCase()}:`);
        values.slice(0, 20).forEach(val => console.log(`      - ${val}`));
        if (values.length > 20) {
          console.log(`      ... and ${values.length - 20} more`);
        }
      }
    });

    // Test scraping a few universities
    console.log('\n\n🎓 Testing University Scraping...\n');
    
    const testUniversities = await page.evaluate(() => {
      const universities = [];
      
      // Try multiple strategies to find university elements
      const strategies = [
        () => Array.from(document.querySelectorAll('.university-card, .uni-card')),
        () => Array.from(document.querySelectorAll('[class*="university"]')),
        () => Array.from(document.querySelectorAll('div[class*="card"]')),
        () => Array.from(document.querySelectorAll('li, div, article')).filter(el => {
          const text = el.textContent || '';
          return text.includes('University') || text.includes('Institute');
        })
      ];

      let found = false;
      for (const strategy of strategies) {
        const elements = strategy();
        if (elements.length > 0) {
          elements.slice(0, 10).forEach(el => {
            const text = el.textContent || '';
            const nameMatch = text.match(/([A-Z][^,\n]+(?:University|Institute)[^,\n]*)/);
            if (nameMatch) {
              universities.push({
                name: nameMatch[1].trim(),
                fullText: text.substring(0, 150),
                html: el.outerHTML.substring(0, 200)
              });
            }
          });
          found = true;
          break;
        }
      }

      return universities;
    });

    console.log(`✅ Found ${testUniversities.length} universities on current page:\n`);
    testUniversities.forEach((uni, idx) => {
      console.log(`   ${idx + 1}. ${uni.name}`);
      console.log(`      Text: ${uni.fullText.substring(0, 80)}...`);
    });

    console.log('\n\n✅ Test Complete!');
    console.log('📝 Review the output above to understand the page structure.');
    console.log('💡 Use this information to adjust the main scraper if needed.\n');

    // Keep browser open for 10 seconds so user can inspect
    console.log('⏳ Keeping browser open for 10 seconds for inspection...');
    await page.waitForTimeout(10000);

  } catch (error) {
    console.error('❌ Test Error:', error);
  } finally {
    await browser.close();
    console.log('\n👋 Browser closed.');
    process.exit(0);
  }
}

// Run test
testHECStructure().catch(console.error);

